web/vtadmin/src/components/routes/topology/Nodes.tsx

```diff
@@ -0,0 +1,103 @@
+import React from 'react';
+import { MarkerType, Node, Edge } from 'react-flow-renderer';
+import { vtctldata } from '../../../proto/vtadmin';
+
+export const generateGraph = (topology: vtctldata.GetTopologyResponse): { nodes: Array<Node>; edges: Array<Edge> } => {
+    const nodes: Array<Node> = [];
+    const edges: Array<Edge> = [];
+
+    let offset = 0;
+    topology.cells.forEach((cell, i) => {
+        const { nodes: childNodes, edges: childEdges } = getNodesAndEdges(cell, cell.name as string, 0, i + offset);
+        nodes.push(...childNodes);
+        edges.push(...childEdges);
+        offset += maxWidth(cell);
+    });
+
+    return {
+        nodes,
+        edges,
+    };
+};
+
+const getNodesAndEdges = (
+    cell: vtctldata.ITopologyCell,
+    path: string,
+    depth: number,
+    width: number
+): { nodes: Array<Node>; edges: Array<Edge> } => {
+    const nodes: Array<Node> = [];
+    const edges: Array<Edge> = [];
+
+    const parentNode: Node = {
+        id: path,
+        position: { y: depth * 100, x: width * 150 },
```

Comment:
Pulling these `position` numbers into vars would be a good excuse to give them names.